File: src/main/scala/com/kotori316/fluidtank/network/SideProxy.java
package com.kotori316.fluidtank.network;

import java.util.Optional;
import java.util.function.Supplier;

import net.minecraft.world.item.Item;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.fml.loading.FMLEnvironment;
import net.minecraftforge.network.NetworkEvent;

public abstract class SideProxy {

    public abstract Optional<Level> getLevel(NetworkEvent.Context context);

    public static boolean isServer(BlockEntity entity) {
        if (entity == null) return false;
        var world = entity.getLevel();
        return world != null && !world.isClientSide();
    }

    public abstract Item.Properties getTankProperties();

    public abstract Item.Properties getReservoirProperties();

    public static SideProxy get() {
        return switch (FMLEnvironment.dist) {
            case CLIENT -> new Client().get();
            case DEDICATED_SERVER -> new Server().get();
        };
    }

    private static class Client implements Supplier<SideProxy> {
        @Override
        @OnlyIn(Dist.CLIENT)
        public SideProxy get() {
            return new ClientProxy();
        }
    }

    private static class Server implements Supplier<SideProxy> {
        @Override
        public SideProxy get() {
            return new ServerProxy();
        }
    }
}

File: src/main/scala/com/kotori316/fluidtank/network/SideProxy.scala
package com.kotori316.fluidtank.network

import java.util.function.Supplier

import net.minecraft.world.item.Item
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraftforge.api.distmarker.{Dist, OnlyIn}
import net.minecraftforge.fml.loading.FMLEnvironment
import net.minecraftforge.network.NetworkEvent

object SideProxy {
  def isServer(entity: BlockEntity): Boolean = {
    if (entity == null) {
      false
    } else {
      val world = entity.getLevel
      world != null && !world.isClientSide
    }
  }

  def get: SideProxy = FMLEnvironment.dist match {
    case Dist.DEDICATED_SERVER => new SideProxy.Server().get
    case Dist.CLIENT => new SideProxy.Client().get
  }

  private class Client extends Supplier[SideProxy] {
    /**
     * Do not forget to specify the return type.
     * [[ClientProxy]] is client only class so it will cause crash of loading in invalid side.
     */
    @OnlyIn(Dist.CLIENT)
    override def get: SideProxy = new ClientProxy
  }

  private class Server extends Supplier[SideProxy] {
    override def get: SideProxy = new ServerProxy
  }
}

abstract class SideProxy {
  def getLevel(context: NetworkEvent.Context): Option[Level]

  def getTankProperties: Item.Properties

  def getReservoirProperties: Item.Properties
}
